cdap-data-fabric/src/main/java/io/cdap/cdap/data2/metadata/lineage/field/FieldLineageInfo.java

```diff
@@ -376,12 +379,17 @@ private void computeIncomingSummaryHelper(EndPointField field, Operation current
 
       // for all the input fields of the previous operation if the origin was current operation (remember we are
       // traversing backward)
-      ReadOperation read = (ReadOperation) currentOperation;
-      EndPoint source = read.getSource();
-      for (InputField inputField : inputFields) {
-        if (inputField.getOrigin().equals(currentOperation.getName())) {
-          sourceEndPointFields.add(new EndPointField(source, inputField.getName()));
+      if (currentOperation.getType() == OperationType.READ) {
+        ReadOperation read = (ReadOperation) currentOperation;
+        EndPoint source = read.getSource();
+        for (InputField inputField : inputFields) {
+          if (inputField.getOrigin().equals(currentOperation.getName())) {
+            sourceEndPointFields.add(new EndPointField(source, inputField.getName()));
+          }
         }
+      } else {
+        // note the source of every generated field is NULL_EPF
+        sourceEndPointFields.add(NULL_EPF);
```

Comment:
The logic here is mostly correct. There is only one problem: if there is a mixture of dropped and generated fields, the NULL_EPF will get removed at https://github.com/cdapio/cdap/blob/38c258f70ad937de67a44030719c6a7554b3c362/cdap-data-fabric/src/main/java/io/cdap/cdap/data2/metadata/lineage/field/FieldLineageInfo.java#L146. And the outgoing summary will have NULL_EPF which we don't need. We might need to change the key of the incomingSummary to an object that can identify the type of the operation.

Reply:
Can we fix this by not removing NULL_EPF if there are generatedFields?